services/placement/filter/mirrored.go

```diff
@@ -0,0 +1,435 @@
+	}
+	return weightToHostsMap, nil
+}
+
+// pairHosts looks at the given hosts and try to make as many pairs as possible.
+func pairHosts(hosts []host, rf int) [][]host {
+	if len(hosts) < rf {
+		// When the number of hosts are less than rf, no pairs can be made.
+		return nil
+	}
+
+	var (
+		uniqRacks = make(map[string]*rack, len(hosts))
+		rh        = &rackHeap{racks: make([]*rack, 0, len(hosts))}
+	)
+	for _, h := range hosts {
+		r, ok := uniqRacks[h.rack]
+		if !ok {
+			r = &rack{
+				rack:  h.rack,
+				hosts: make([]host, 0, rf),
+			}
+
+			uniqRacks[h.rack] = r
+			rh.racks = append(rh.racks, r)
+		}
+		r.hosts = append(r.hosts, h)
+	}
+
+	heap.Init(rh)
+
+	res := make([][]host, 0, len(hosts))
+	for rh.Len() >= rf {
+		// When there are more than rf racks availabl, try to make a pair.
+		seenRacks := make(map[string]*rack, rf)
+		pairs := make([]host, 0, rf)
+		for i := 0; i < rf; i++ {
+			var r *rack
+			for rh.Len() > 0 {
+				// Always prefer to take from the largest rack.
+				r = rh.Pop().(*rack)
+				if _, ok := seenRacks[r.rack]; !ok {
+					seenRacks[r.rack] = r
+					break
+				}
+			}
+			if r == nil {
+				break
+			}
+			// Move the host from the rack to the pair.
+			pairs = append(pairs, r.hosts[len(r.hosts)-1])
+			r.hosts = r.hosts[:len(r.hosts)-1]
+		}
+		if len(pairs) == rf {
+			res = append(res, pairs)
+		}
+		for _, r := range seenRacks {
+			if len(r.hosts) > 0 {
+				rh.Push(r)
+			}
+		}
+	}
+
+	return res
+}
+
+func pairInstances(hostPairs [][]host) ([][]services.PlacementInstance, error) {
+	var instancePairs = make([][]services.PlacementInstance, 0, len(hostPairs))
+	for _, hostPair := range hostPairs {
+		for port, instance := range hostPair[0].portToInstance {
+			instancePair := make([]services.PlacementInstance, 0, len(hostPair))
+			instancePair = append(instancePair, instance)
+			for _, otherHost := range hostPair[1:] {
+				otherInstance, ok := otherHost.portToInstance[port]
+				if !ok {
+					return nil, fmt.Errorf("could not find port %s on host %s", port, otherHost.name)
+				}
+				instancePair = append(instancePair, otherInstance)
+			}
+			instancePairs = append(instancePairs, instancePair)
+		}
+	}
+	return instancePairs, nil
+}
+
+func assignShardsToPairedInstances(
+	pairs [][]services.PlacementInstance,
+	p services.Placement,
+) []services.PlacementInstance {
+	var (
+		instances   = make([]services.PlacementInstance, 0, p.ReplicaFactor()*len(pairs))
+		shardSetIDs = nextNShardSetIDs(p, len(pairs))
+	)
+	for i, pair := range pairs {
+		ssID := shardSetIDs[i]
+		for _, instance := range pair {
+			instances = append(instances, instance.SetShardSetID(ssID))
+		}
+	}
+
+	return instances
+}
+
+func nextNShardSetIDs(p services.Placement, n int) []string {
+	var (
+		instances          = p.Instances()
+		currentShardSetIDs = make(map[int]struct{}, len(instances))
+	)
+	for _, instance := range instances {
+		ssID, err := strconv.Atoi(instance.ShardSetID())
```

Comment:
The `shardSetIDAfter(ssid string)` would simply be passed in from the options and has no default implementation. Actually you could have a `ShardSetIDGenerator` interface with a `NewShardSetIDGenerator(startID string)` constructor and a `NextID() (string, error)` method that simply returns the next id, which might be cleaner. This way the generator can generate uuids if users prefer to.

Don't think it's a big concern for not reusing shard set ids, given that it could be arbitrary strings.